### PR TITLE
acl: Add support for globbing namespaces

### PR DESCRIPTION
**Pull request by dantoml**
This adds basic support for globbing namespaces in acl
definitions.

For concrete definitions, we merge all of the defined policies at load time, and
perform a simple lookup later on. If an exact match of a concrete
definition is found, we do not attempt to resolve globs.

For glob definitions, we merge definitions of exact replicas of a glob.

When loading a policy for a glob defintion, we choose the glob that has
the closest match to the namespace we are resolving for. We define the
closest match as the one with the _smallest character difference_
between the glob and the namespace we are matching.

### CODE REVIEW

File: acl/acl_test.go [hunk 1]
```diff
@@ -262,3 +262,131 @@ func TestAllowNamespace(t *testing.T) {
...
+		NS         string
+		Difference int
+	}{
+		{
+			Policy:     `namespace "production-*" { policy = "write" }`,
+			NS:         "production-api",
+			Difference: 3,
+		},
+		{
+			Policy:     `namespace "production-*" { policy = "write" }`,
```
[the remark this answers is not in the document]
Comment (dadgar) [reply acl/acl_test.go]:
```
Policy: `namespace "p*-*" { policy = "write" }`
NS: "production-api"
Difference: x
```

File: acl/acl.go [hunk 2]
```diff
@@ -44,6 +47,10 @@ type ACL struct {
 	// namespaces maps a namespace to a capabilitySet
 	namespaces *iradix.Tree
 
+	// wildcardNamespaces maps a glob pattern of a namespace to a capabilitySet
+	// We use an iradix for the purposes of ordered iteration.
+	wildcardNamespaces *iradix.Tree
```
Comment (dadgar) [inline acl/acl.go]: Can we add documentation changes to this PR as well: 
https://github.com/hashicorp/nomad/blob/master/website/source/guides/security/acl.html.markdown

Comment (dantoml) [reply acl/acl.go]: Done - I'm not _too_ happy with my draft on those tho, so would love a second look from you / @Omar-Khawaja.

Comment (Omar-Khawaja) [reply acl/acl.go]: @dantoml lgtm. I would probably change up the wording very slightly.

Namespace definitions may also include globs which can be used to provide access to a category of namespaces while allowing more specific control over privileges where required. For example, the below policy allows read access to most production namespaces but allows write access to the "production-api" namespace while denying access to the "production-ppi" namespace.

File: website/source/guides/security/acl.html.markdown [hunk 3]
```diff
@@ -253,6 +253,36 @@ namespace "default" {
 }
 ```
 
+Namespace definitions may also include globs, that can be used to provide access to a category of namespaces, while allowing more specific privileges where required. For example, the below policy allows read access to most production namespaces, but allows write access to the `"production-api"` namespace, and rejects any access to the `"production-ppi"` namespace.
+
```
Comment (dadgar) [inline website/source/guides/security/acl.html.markdown]: Namespaces definitions may also include globs, allowing a single policy definition to apply to a set of namespaces. For example, the below policy allows read access to most production namespaces, but allows write access to the `"production-api"` namespace, and rejects any access to the `"production-web"` namespace.

File: website/source/guides/security/acl.html.markdown [hunk 4]
```diff
@@ -253,6 +253,36 @@ namespace "default" {
...
+```
+namespace "production-*" {
+    policy = "read"
+}
+
+namespace "production-api" {
+    policy = "write"
+}
+
+namespace "production-ppi" {
```
Comment (dadgar) [inline website/source/guides/security/acl.html.markdown]: Lets use web (or something similar); not sure what ppi is and may confuse people

File: website/source/guides/security/acl.html.markdown [hunk 5]
```diff
@@ -253,6 +253,36 @@ namespace "default" {
...
+    policy = "deny"
+}
+```
+
+Namespaces are matched to their policies first by performing a lookup on any _exact match_, before falling back to performing a glob based lookup. When looking up namespaces by glob, the matching policy with the fewest number of matched characters will be chosen. For example:
```
Comment (dadgar) [inline website/source/guides/security/acl.html.markdown]: fewest -> greatest

File: website/source/guides/security/acl.html.markdown [hunk 6]
```diff
@@ -253,6 +253,36 @@ namespace "default" {
...
+
+```
+namespace "*-ppi" {
```
Comment (dadgar) [inline website/source/guides/security/acl.html.markdown]: Change to web as well

**Comment by dadgar** [comment]
Don't forget to add to changelog

**Pull request merged**